node/src/components/block_synchronizer/peer_list.rs

```diff
@@ -148,34 +145,40 @@ impl PeerList {
         PeersStatus::Sufficient
     }
 
-    pub(super) fn qualified_peers(&self, rng: &mut NodeRng) -> Vec<NodeId> {
-        let up_to = self.max_simultaneous_peers as usize;
-
-        // get most useful up to limit
-        let mut peers: Vec<NodeId> = self
-            .peer_list
+    fn get_peers_by_quality(
+        &self,
+        rng: &mut NodeRng,
+        up_to: usize,
+        peer_quality: PeerQuality,
+    ) -> Vec<NodeId> {
+        self.peer_list
             .iter()
-            .filter(|(_peer, quality)| **quality == PeerQuality::Reliable)
+            .filter(|(_peer, quality)| **quality == peer_quality)
             .choose_multiple(rng, up_to)
             .into_iter()
             .map(|(peer, _)| *peer)
-            .collect();
+            .collect()
+    }
 
-        // if below limit get semi-useful
+    pub(super) fn qualified_peers(&self, rng: &mut NodeRng) -> Vec<NodeId> {
+        let up_to = self.max_simultaneous_peers as usize;
+
+        // get most useful up to limit
+        let mut peers = self.get_peers_by_quality(rng, up_to, PeerQuality::Reliable);
+
+        // if below limit get unknown peers which may or may not be useful
         let missing = up_to.saturating_sub(peers.len());
         if missing > 0 {
-            let better_than_nothing = self
-                .peer_list
-                .iter()
-                .filter(|(_peer, quality)| {
-                    **quality == PeerQuality::Unreliable || **quality == PeerQuality::Unknown
-                })
-                .choose_multiple(rng, missing)
-                .into_iter()
-                .map(|(peer, _)| *peer);
-
-            peers.extend(better_than_nothing);
+            peers.extend(self.get_peers_by_quality(rng, up_to, PeerQuality::Unknown));
         }
+
+        // if still below limit try unreliable peers again until we have the chance to refresh the
+        // peer list
+        let missing = up_to.saturating_sub(peers.len());
+        if missing > 0 {
+            peers.extend(self.get_peers_by_quality(rng, up_to, PeerQuality::Unreliable));
```

Comment:
Should we also try to get peers of `Unknown` quality here?

Reply:
Yes, the intention was to first get `Reliable` peers, then if those are not enough extend the peers list and get `Unknown` peers and even if then we don't have enough peers, get unreliable ones. But I did a mistake here and used `up_to` in all of the calls instead of using `missing`. Good catch! I'll fix this.